src/Hash.js

```diff
@@ -10,41 +10,50 @@ import { smi } from './Math';
 const defaultValueOf = Object.prototype.valueOf;
 
 export function hash(o) {
-  switch (typeof o) {
+  if (o == null) {
+    return hashNullish(o);
+  }
+
+  if (typeof o.hashCode === 'function') {
+    // Drop any high bits from accidentally long hash codes.
+    return smi(o.hashCode(o));
+  }
+
+  const v = valueOf(o);
+
+  if (v == null) {
+    return hashNullish(v);
+  }
```

Comment:
Can you explain why we are moving this code of out of the `'object' || 'function'` case?

Reply:
Quite simply: `moment().valueOf()`, which returns a number (the timestamp). The current code assumes incorrectly that the result of `object.valueOf()` is always an object.

Reply:
Oh, I see now you're tackling https://github.com/immutable-js/immutable-js/issues/1643 as well in this PR. I was trying to figure out what this code had to do with symbols, but now it makes sense.